### PR TITLE
UCP/EP: Consider all memtypes for GET_ZCOPY lanes

### PR DESCRIPTION
**Pull request by bureddy**
## Why ?
PR #4912, RMA bw lanes for some mem type are ignored when there is another mem type lane is 10x (UCX_MULTI_LANE_MAX_RATIO) faster 
For example. HOST lanes ignored when PR https://github.com/openucx/ucx/pull/5344 set cuda_ipc bw with max NVlanes

### CODE REVIEW

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
http://hpc-master.lab.mtl.com:8080/blue/rest/organizations/jenkins/pipelines/ucx/runs/5892/nodes/131/steps/898/log/?start=0
```
[2020-07-06T08:31:23.236Z] Making all in src/ucp
[2020-07-06T08:31:23.236Z] make[2]: Entering directory `/scrap/jenkins/workspace/ucx-8/build-test/src/ucp'
[2020-07-06T08:31:23.236Z]   CC       core/libucp_la-ucp_context.lo
[2020-07-06T08:31:23.236Z]   CC       core/libucp_la-ucp_am.lo
[2020-07-06T08:31:23.236Z]   CC       core/libucp_la-ucp_ep.lo
[2020-07-06T08:31:23.236Z]   CC       core/libucp_la-ucp_listener.lo
[2020-07-06T08:31:23.491Z] /scrap/jenkins/workspace/ucx-8/contrib/../src/ucp/core/ucp_ep.c:1474:13: error: comparison of constant 64 with expression of type 'ucs_memory_type_t' (aka 'enum ucs_memory_type') is always true [-Werror,-Wtautological-constant-out-of-range-compare]
[2020-07-06T08:31:23.491Z]             ucs_for_each_bit(mem_type, md_attr->cap.reg_mem_types) {
[2020-07-06T08:31:23.491Z]             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2020-07-06T08:31:23.491Z] /scrap/jenkins/workspace/ucx-8/contrib/../src/ucs/sys/math.h:151:52: note: expanded from macro 'ucs_for_each_bit'
[2020-07-06T08:31:23.491Z]     for ((_index) = ucs_ffs64_safe(_map); (_index) < 64; \
[2020-07-06T08:31:23.491Z]                                           ~~~~~~~~ ^ ~~
[2020-07-06T08:31:23.491Z] /scrap/jenkins/workspace/ucx-8/contrib/../src/ucp/core/ucp_ep.c:1492:17: error: comparison of constant 64 with expression of type 'ucs_memory_type_t' (aka 'enum ucs_memory_type') is always true [-Werror,-Wtautological-constant-out-of-range-compare]
[2020-07-06T08:31:23.491Z]                 ucs_for_each_bit(mem_type, md_attr->cap.reg_mem_types) {
[2020-07-06T08:31:23.491Z]                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2020-07-06T08:31:23.491Z] /scrap/jenkins/workspace/ucx-8/contrib/../src/ucs/sys/math.h:151:52: note: expanded from macro 'ucs_for_each_bit'
[2020-07-06T08:31:23.491Z]     for ((_index) = ucs_ffs64_safe(_map); (_index) < 64; \
[2020-07-06T08:31:23.491Z]                                           ~~~~~~~~ ^ ~~
[2020-07-06T08:31:23.491Z] 2 errors generated.
[2020-07-06T08:31:23.491Z] make[2]: *** [core/libucp_la-ucp_ep.lo] Error 1
[2020-07-06T08:31:23.491Z] make[2]: *** Waiting for unfinished jobs....
[2020-07-06T08:31:24.884Z] make[2]: Leaving directory `/scrap/jenkins/workspace/ucx-8/build-test/src/ucp'
[2020-07-06T08:31:24.884Z] make[1]: *** [all-recursive] Error 1
[2020-07-06T08:31:24.884Z] make[1]: Leaving directory `/scrap/jenkins/workspace/ucx-8/build-test'
[2020-07-06T08:31:24.884Z] make: *** [all] Error 
```

File: src/ucp/core/ucp_ep.c [hunk 1]
```diff
@@ -1477,23 +1486,28 @@ ucs_status_t ucp_ep_config_init(ucp_worker_h worker, ucp_ep_config_t *config,
...
             iface_attr = ucp_worker_iface_get_attr(worker, rsc_index);
+            md_attr    = &context->tl_mds[context->tl_rscs[rsc_index].md_index].attr;
 
             /* GET Zcopy */
             if (iface_attr->cap.flags & UCT_IFACE_FLAG_GET_ZCOPY) {
-                scale = ucp_tl_iface_bandwidth(context, &iface_attr->bandwidth) /
-                        rndv_max_bw;
-                if (scale < (1. / context->config.ext.multi_lane_max_ratio)) {
-                    continue;
+                ucs_for_each_bit(mem_type_index, md_attr->cap.reg_mem_types) {
```
Comment (dmitrygx) [inline src/ucp/core/ucp_ep.c:1493]: I don't think this is correct - in this for_each loop we set `config->tag.rndv.scale[lane]` and `config->tag.rndv.get_zcopy_lanes[get_zcopy_lane_count++]` values several times = # of memory type MDs.
so, it makes `get_zcopy_lane_count` greater than we really have

should we have `min_get_zcopy`/`max_get_zcopy`/`get_zcopy_lanes`/`scale` for each memory type? then it should work

Comment (bureddy) [reply src/ucp/core/ucp_ep.c:1493]: i have `"break"` at the line :1509 . the lane will be added only one. am I missing something?

Comment (dmitrygx) [reply src/ucp/core/ucp_ep.c:1493]: @bureddy thanks, I see now. didn't notice this `break` operator
does it mean if we have two lanes configured for intra-node RNDV(`knem`, `cuda_ipc`), before this fix `knem` wasn't selected due to low performance in comparison to `cuda_ipc`? so, we didn't perform RNDV for HOST memory buffers.

File: src/ucp/core/ucp_ep.c [hunk 2]
```diff
@@ -1477,23 +1486,28 @@ ucs_status_t ucp_ep_config_init(ucp_worker_h worker, ucp_ep_config_t *config,
...
+                    ucs_assert(mem_type_index < UCS_MEMORY_TYPE_LAST);
+                    scale = ucp_tl_iface_bandwidth(context, &iface_attr->bandwidth) /
+                            rndv_max_bw[mem_type_index];
+                    if (scale < (1. / context->config.ext.multi_lane_max_ratio)) {
+                        continue;
+                    }
+
+                    config->tag.rndv.min_get_zcopy = ucs_max(config->tag.rndv.min_get_zcopy,
```
Comment (dmitrygx) [inline src/ucp/core/ucp_ep.c:1501]:
```C
config->tag.rndv.min_get_zcopy = ucs_max(config->tag.rndv.min_get_zcopy,
                                         iface_attr->cap.get.min_zcopy);
config->tag.rndv.max_get_zcopy = ucs_min(config->tag.rndv.max_get_zcopy,
                                         iface_attr->cap.get.max_zcopy);

ucs_assert(get_zcopy_lane_count < UCP_MAX_LANES);
config->tag.rndv.get_zcopy_lanes[get_zcopy_lane_count++] = lane;
config->tag.rndv.scale[lane]                             = scale;
break;
```

File: src/ucp/core/ucp_ep.c [hunk 3]
```diff
@@ -1461,12 +1466,16 @@ ucs_status_t ucp_ep_config_init(ucp_worker_h worker, ucp_ep_config_t *config,
             continue;
         }
 
+        md_attr    = &context->tl_mds[context->tl_rscs[rsc_index].md_index].attr;
         iface_attr = ucp_worker_iface_get_attr(worker, rsc_index);
         if (iface_attr->cap.flags & UCT_IFACE_FLAG_GET_ZCOPY) {
             /* only GET Zcopy RNDV scheme supports multi-rail */
             bw          = ucp_tl_iface_bandwidth(context,
```
Comment (dmitrygx) [inline src/ucp/core/ucp_ep.c:1473]: remove exceeded whitespaces, they are not needed anymore

File: src/ucp/core/ucp_ep.c (hunk 1 above)
Comment (bureddy) [reply src/ucp/core/ucp_ep.c:1493]: correct

**Comment by bureddy** [comment]
@yosefe can you please review

**Pull request merged**